fs-widget/fs-widget.lua

```diff
@@ -6,28 +6,74 @@ local gears = require("gears")
 
 local storage_bar_widget = {}
 
+--- Table with widget configuration, consists of three sections:
+---  - general - general configuration
+---  - widget - configuration of the widget displayed on the wibar
+---  - popup - configuration of the popup
+local config = {}
+
+-- general
+config.mounts = { '/' }
+config.refresh_rate = 60
+
+-- wibar widget
+config.widget_width = 40
+config.widget_bar_color = '#aaaaaa'
+config.widget_onclick_bg = '#ff0000'
+config.widget_border_color = '#535d6c66'
+config.widget_background_color = '#22222233'
+
+-- popup
+config.popup_bg = '#22222233'
+config.popup_border_width = 1
+config.popup_border_color = '#535d6c66'
+config.popup_bar_color = '#aaaaaa'
+config.popup_bar_background_color = '#22222233'
+config.popup_bar_border_color = '#535d6c66'
+
 local function worker(user_args)
     local args = user_args or {}
-    local mounts = args.mounts or {'/'}
-    local timeout = args.timeout or 60
+
+    if beautiful ~= nil then
```

Comment:
Is this working? You require the beautiful library, so the beautiful variable is the instance. It has to exist. 

Maybe you should replace this check by :

```lua
if not #beautiful.get() then 
```

Or another more robust check would be to test each field you need from beautiful. 

(Also, I personally don't think this is a good implementation. You basically replace the module level `config` values from the instance constructor. I mean, it works but it feels like encapsulation breakage since an instance is altering a class level property that can affect other instances.)

Reply:
This is nice example when the unit tests can detect fast these errors. I know that this is lua, not java, or C#, but unit tests are necessary for every language. It is larger investment from the beginning, but this investment will return very fast.

Reply:
> Or another more robust check would be to test each field you need from beautiful.

Like this?

```lua
config.widget_bar_color = beautiful.fg_normal == nil and config.widget_bar_color or beautiful.fg_normal
```

> (Also, I personally don't think this is a good implementation. You basically replace the module level `config` values from the instance constructor. I mean, it works but it feels like encapsulation breakage since an instance is altering a class level property that can affect other instances.)

You're right, maybe it would be better to shadow `config` in the constructor then? Something like this:

```lua
local function worker(user_args)
    local args = user_args or {}
    loca config = config

    if #beautiful.get() ~= nil then
        config.widget_bar_color = beautiful.fg_normal == nil and config.widget_bar_color or beautiful.fg_normal
```

What do you think?

Reply:
> > Or another more robust check would be to test each field you need from beautiful.
> 
> Like this?
> 
> ```lua
> config.widget_bar_color = beautiful.fg_normal == nil and config.widget_bar_color or beautiful.fg_normal
> ```

It can even be easier with:

```lua
config.widget_bar_color = config.widget_bar_color or beautiful.fg_normal
```

The Lua `or` operator returns the first "true" value. SO the nullity check is handled by `or`.

> > (Also, I personally don't think this is a good implementation. You basically replace the module level `config` values from the instance constructor. I mean, it works but it feels like encapsulation breakage since an instance is altering a class level property that can affect other instances.)
> 
> You're right, maybe it would be better to shadow `config` in the constructor then? Something like this:
> 
> ```lua
> local function worker(user_args)
>     local args = user_args or {}
>     loca config = config
> 
>     if #beautiful.get() ~= nil then
>         config.widget_bar_color = beautiful.fg_normal == nil and config.widget_bar_color or beautiful.fg_normal
> ```
> 
> What do you think?

I'm not really a fan of shadowing values. I don't even think it would work as expected here, `beautiful.fg_normal == nil and config.widget_bar_color` I think it will use the "new" config table instead of the global.

I did a quick PR to show you what I have in mind with this pattern.